connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java

```diff
@@ -213,7 +219,12 @@ public synchronized void putConnectorConfig(String connName,
                 return;
             }
 
-            updateConnectorTasks(connName);
+            final Map<String, String> clientAndConverterConfigs = clientAndConverterConfigs(config);
+            final Map<String, String> priorClientAndConverterConfigs =
+                connectorClientAndConverterConfigs.put(connName, clientAndConverterConfigs);
+            boolean forceReconfigure = !clientAndConverterConfigs.equals(priorClientAndConverterConfigs);
```

Comment:
Rather than cache the previous connector configs, could we not just modify this method a bit earlier to compare the existing (transformed) connector config  to the (transformed) supplied connector config? Don't the (transformed) connector configurations stored in the `configBackingStore` (and thus accessible in the `configState`) contain all of the client overrides and converter settings? If so, comparing the transformed versions of those is actually great, because we'd also detect the tasks need to be updated even when the user-supplied connector config didn't change but the *resolved variable values* do change.

For example, this method would essentially become:
```
            boolean created = false;
            boolean changed = false;     // <--- added
            if (configState.contains(connName)) {
                if (!allowReplace) {
                    callback.onCompletion(new AlreadyExistsException("Connector " + connName + " already exists"), null);
                    return;
                }
                // Compare the existing transformed connector config to the new transformed config
                // and if anything changed (including just resolved secrets) then we should force an update to the task configs
                Map<String, String> existingXformedConnectorConfig = configState.connectorConfig(connName);
                Map<String, String> newXformedConnectorConfig = worker.configTransformer().transform(config);
                changed = !existingXformedConnectorConfig.equals(newXformedConnectorConfig);

                worker.stopConnector(connName);
            } else {
                created = true;
            }

            configBackingStore.putConnectorConfig(connName, config);   // <-- same as your suggestion

            if (!startConnector(connName)) {
                callback.onCompletion(new ConnectException("Failed to start connector: " + connName), null);
                return;
            }

            updateConnectorTasks(connName, changed);
```

WDYT?